### ISSUE TITLE
REST API purpose

### ISSUE THREAD

**Issue by bryansippel** [christianjianelli/scnblog8#1]
I downloaded and installed this stuff and just need to confirm what it will enable me to do. Is this to be used from SAP ABAP to consume an external REST service or is it intended to host a REST service from SAP ABAP and provided methods for external customers to interact with data on our system?

**Comment by christianjianelli** [comment]
The second one. Its purpose is to expose a REST API on your SAP WebAS ABAP to be consumed from external client applications.

**Comment by bryansippel** [comment]
Bummer. I was hoping it would assist me in formatting JSON to be used to consume an external REST service.

I am having a terrible time locating specific examples of how to use CL_HTTP_CLIENT via ABAP (SAPGui, not Eclipse). We are not on 7.4 so we don't have the new JSON class.

I loaded ZJSON, but I still can't find someone who can simply tell me if calling the CL_HTTP_CLIENT->REQUEST->SET_CDATA method will attach my JSON formatted request text to the "payload" that is expected by the external REST service.

I can't begin to count the number of links I have clicked on, many referring to 7.4, many others showing only Java or XML, but no ABAP using CL_HTTP_CLIENT send/receive specifically with SET_DATA or SET_CDATA with JSON formatted text.

I have submitted a question on SDN, but I'm running out of ideas.

Any suggestions you could provide would be so very greatly appreciated. Thanks.

**Comment by christianjianelli** [comment]
Check the example report bellow

_&---------------------------------------------------------------------_
_& Report  ZR_HTTP_CLIENT
*&
*&---------------------------------------------------------------------_
_&
*&
*&---------------------------------------------------------------------_

REPORT  zr_http_client.

PARAMETERS: p_url  TYPE url
              DEFAULT 'http://www.example.com'
              LOWER CASE.

" Objects
DATA: o_client TYPE REF TO if_http_client.

" Elementary
DATA: url TYPE string,
      msg TYPE string,
      request_cdata TYPE string,
      response_cdata TYPE string.

---
- START-OF-SELECTION                                                   *
  
  ---
  
  START-OF-SELECTION.
  
  url = p_url.
  
  cl_http_client=>create_by_url(
    EXPORTING
      url                = url
    IMPORTING
      client             = o_client
    EXCEPTIONS
      argument_not_found = 1
      plugin_not_active  = 2
      internal_error     = 3
      OTHERS             = 4
  ).
  
  IF sy-subrc <> 0.
    o_client->get_last_error( IMPORTING message = msg ).
    WRITE: / msg.
    STOP.
  ENDIF.
  
  o_client->request->set_method( 'GET' ). " GET, POST, PUT, DELETE
  
  request_cdata = '{"success" :"true","code" :200,"msg" :"Hello World!"}'.
  
  o_client->request->set_cdata( request_cdata ).
  
  o_client->send(
    EXCEPTIONS
      http_communication_failure = 1
      http_invalid_state         = 2
      http_processing_failed     = 3
      http_invalid_timeout       = 4
      OTHERS                     = 5
  ).
  
  IF sy-subrc <> 0.
    o_client->get_last_error( IMPORTING message = msg ).
    WRITE: / msg.
    STOP.
  ENDIF.
  
  o_client->receive(
    EXCEPTIONS
      http_communication_failure = 1
      http_invalid_state         = 2
      http_processing_failed     = 3
      OTHERS                     = 4
  ).
  
  response_cdata = o_client->response->get_cdata( ).

**Comment by bryansippel** [comment]
Thank you very much for your post. It helps to confirm what I am already doing although I am also doing a refresh_request and setting some HTML header parameters before calling the send method.

I had tried hardcoding a JSON text to confirm that process and it appeared to work properly, but what I really needed was an example showing which ZJSON methods should be called with their associated parameters at the appropriate points in the process in order to translate the ABAP stuff into JSON stuff before the SEND and vice-versa with the response string returned from the REST service.

After playing around with it for a while, I discovered the zcl_json_document=>create_with_data method can be used to import the ABAP structure content to the JSON document class and the get_json method can be used to extract it into a string to pass into the SET_CDATA method.

After calling the SEND, RECEIVE, and GET_CDATA methods, I also discovered the create_with_json method to load the JSON text into the JSON document class and the get_data method to convert it to my ABAP deep structures.

Thanks so much for your input!!
